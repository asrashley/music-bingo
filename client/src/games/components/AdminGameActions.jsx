import React from 'react';
import PropTypes from 'prop-types';
import { saveAs } from 'file-saver';

import {
  BusyDialog, ConfirmDialog, FileDialog, ProgressDialog
} from '../../components';

import { addMessage } from '../../messages/messagesSlice';
import { importDatabase } from '../../admin/adminSlice';
import {
  deleteGame, fetchGamesIfNeeded, importGame, invalidateGames
} from '../gamesSlice';

import { api } from '../../endpoints';

import { GamePropType } from '../types/Game';

export function AdminActionPanel({ deleteGame, exportGame, importGame, game }) {
  return (
    <div className="action-panel">
      <button className="btn btn-primary ml-2" onClick={importGame}>
        Import a game
      </button>
      {(game && exportGame) && <button className="btn btn-primary ml-2"
        onClick={exportGame}>Export game</button>}
      {(game && deleteGame) && <button className="btn btn-danger ml-2"
        onClick={deleteGame}>Delete game</button>}
    </div>
  );
}

AdminActionPanel.propTypes = {
  game: GamePropType,
  deleteGame: PropTypes.func,
  exportGame: PropTypes.func,
  importGame: PropTypes.func.isRequired,
};

function ErrorMessage({ error }) {
  if (!error) {
    return null;
  }
  return (
    <div className="alert alert-warning" role="alert">
      <span className="error-message">{error}</span>
    </div>
    );
}
ErrorMessage.propTypes = {
  error: PropTypes.string
};

export class AdminGameActions extends React.Component {
  static DATABASE_IMPORT = 1;
  static GAME_IMPORT = 2;

  static propTypes = {
    databaseImporting: PropTypes.object,
    dispatch: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
    game: GamePropType,
    gameImporting: PropTypes.object,
    importing: PropTypes.object,
  };

  state = {
    ActiveDialog: null,
    dialogData: null,
    importType: 0,
    error: null
  };

  render() {
    const { ActiveDialog, dialogData, error } = this.state;
    const { children, game, importing } = this.props;

    return (
      <React.Fragment>
        <ErrorMessage error={error} />
        {children}
        <AdminActionPanel
          game={game}
          deleteGame={this.confirmDelete}
          exportGame={this.exportGame}
          importGame={this.onClickImportGame}
        />;
        {ActiveDialog && <ActiveDialog backdrop {...dialogData} {...importing} /> }
      </React.Fragment>);
  }

  confirmDelete = (ev) => {
    ev.preventDefault();
    const { game } = this.props;
    this.setState({
      ActiveDialog: ConfirmDialog,
      dialogData: {
        changes: [
          `Delete group ${game.id}`
        ],
        title: "Confirm delete game",
        onCancel: this.cancelDialog,
        onConfirm: this.deleteGame,
      }
    });
    return false;
  };

  cancelDialog = () => {
    this.setState({
      ActiveDialog: null,
      dialogData: null,
      importType: 0,
    });
  };

  deleteGame = () => {
    const { game, dispatch, onDelete } = this.props;
    this.setState({ ActiveDialog: null, dialogData: null });
    dispatch(deleteGame(game)).then(result => {
      const { payload } = result;
      if (payload?.error || result.error) {
        this.setState({ error: (payload.error ? payload.error : result.error) });
      } else {
        onDelete(game);
      }
    });
  };

  onClickImportDatabase = () => {
    this.setState({
      ActiveDialog: FileDialog,
      dialogData: {
        title: 'Select a json database file to import',
        accept: '.json,application/json',
        submit: 'Import database',
        onCancel: this.cancelDialog,
        onFileUpload: this.onFileSelected
      },
      importType: AdminGameActions.DATABASE_IMPORT
    });
  };

  onChangeReplaceDate = (e) => {
    const { dialogData } = this.state;
    this.setState({
      dialogData: {
        ...dialogData,
        replaceDate: e.target.checked
      }
    });
  }

  onClickImportGame = () => {
    const extraFields = <div className="form-control">
      <label htmlFor="replaceDate" className="replace-date-label">Replace game start date during import?</label>
      <input type="checkbox" name="replaceDate" defaultChecked onChange={this.onChangeReplaceDate}></input>
    </div>;

    this.setState({
      ActiveDialog: FileDialog,
      dialogData: {
        title: 'Select a gameTracks.json file to import',
        accept: '.json,application/json',
        submit: 'Import game',
        onCancel: this.cancelDialog,
        onFileUpload: this.onFileSelected,
        replaceDate: true,
        extraFields
      },
      importType: AdminGameActions.GAME_IMPORT
    });
  };

  importComplete = () => {
    const { databaseImporting, gameImporting, dispatch } = this.props;
    const { importType } = this.state;
    if (importType === AdminGameActions.DATABASE_IMPORT && databaseImporting?.done === true) {
        document.location.reload();
        return;
    }
    if (importType === AdminGameActions.GAME_IMPORT && gameImporting?.done === true) {
      dispatch(invalidateGames());
      dispatch(fetchGamesIfNeeded());
    }
    this.setState({
      ActiveDialog: null,
      dialogData: null,
      importType: 0,
    });
  };

  exportDatabase = () => {
    const { dispatch } = this.props;
    this.setState(() => ({
      ActiveDialog: BusyDialog,
      dialogData: {
        onClose: this.cancelDialog,
        title: "Exporting database",
        text: "Please wait, exporting database...",
      }
    }), () => {
      dispatch(api.exportDatabase()).then(response => {
        const filename = `database-${Date.now()}.json`;
        return response.payload.blob().then(blob => {
          this.setState({ ActiveDialog: null, dialogData: null });
          saveAs(blob, filename);
          return filename;
        });
      });
    });
  };

  exportGame = () => {
    const { dispatch, game } = this.props;
    dispatch(api.exportGame({
      gamePk: game.pk
    })).then(response => {
      const filename = `game-${game.id}.json`;
      return response.payload.blob().then(blob => {
        saveAs(blob, filename);
        return filename;
      });
    });
  };

  onFileSelected = (file) => {
    var reader = new FileReader();
    reader.onload = (ev) => this.onFileLoaded(file.name, ev);
    reader.onerror = (err) => {
      this.setState({ activeDialog: null, activeDialogData: null });
      addMessage({ type: "error", text: err });
    };
    reader.readAsText(file);
  };

  onFileLoaded(filename, event) {
    const { dispatch } = this.props;
    const { importType } = this.state;
    const { replaceDate } = this.state.dialogData;
    const title = importType === AdminGameActions.GAME_IMPORT ?
      `Importing game from "${filename}"` :
      `Importing database from "${filename}"`;

    this.setState({
      ActiveDialog: ProgressDialog,
      dialogData: {
        title,
        onCancel: this.cancelDialog,
        onClose: this.importComplete
      }
    });
    const data = JSON.parse(event.target.result);
    if (importType === AdminGameActions.GAME_IMPORT) {
      if (replaceDate === true && 'Games' in data) {
        const start = new Date().toISOString();
        const end = new Date(Date.now() + 12*3600000).toISOString();
        data.Games = data.Games.map(game => ({
          ...game,
          start,
          end
        }));
      }
      dispatch(importGame(filename, data));
    } else {
      dispatch(importDatabase(filename, data));
    }
  }
}


