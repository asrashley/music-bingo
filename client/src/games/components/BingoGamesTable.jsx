import React from 'react';
import { reverse } from 'named-urls';
import routes from '../../routes';
import { Link } from 'react-router-dom';

function formatTime(value) {
  let hours = value.getHours();
  const ampm = value.getHours() >= 12 ? 'pm' : 'am';
  hours = hours % 12;
  if (hours === 0) {
    hours = 12;
  }
  //hours = `0${hours}`.slice(-2);
  const minutes = `0${value.getMinutes()}`.slice(-2);
  return `${hours}:${minutes} ${ampm}`;
}

const TableRow = ({ game, past }) => {
  let ticketUrl, themeUrl, linkText;

  if (past) {
    linkText = 'View track listing';
    ticketUrl = themeUrl = reverse(`${routes.trackListing}`, { gameId: game.id });
  } else if (game.userCount === 0) {
    linkText = "Choose tickets";
    ticketUrl = reverse(`${routes.game}`, { gameId: game.id });
    themeUrl = reverse(`${routes.game}`, { gameId: game.id });
  } else {
    ticketUrl = reverse(`${routes.play}`, { gameId: game.id });
    themeUrl = reverse(`${routes.game}`, { gameId: game.id });
    linkText = `You have chosen ${game.userCount} ticket${(game.userCount > 1) ? 's' : ''}`;
  }
  const start = new Date(game.start);
  return (
    <React.Fragment>
      {game.round === 1 && (<tr>
        <th colSpan="4">{start.toDateString()}</th>
      </tr>)}
      <tr className={`round-${game.round}`}>
        <td className="round-column">{game.round}</td>
        <td className="date-column">{formatTime(start)}</td>
        <td className="theme-column"><Link to={themeUrl}>{game.title}</Link></td>
        <td className="ticket-column"><Link to={ticketUrl}>{linkText}</Link></td>
      </tr>
    </React.Fragment>
  );
};

export const BingoGamesTable = ({ title, games, past, onReload }) => {
  return (
    <table className="table table-striped table-bordered game-list">
      <thead>
        <tr>
          <th colSpan="4" className="available-games">{title}
            <button className="btn btn-light refresh-icon btn-sm" onClick={onReload}>&#x21bb;</button>
          </th>
        </tr>
        <tr>
          <th className="round-column">Round</th>
          <th className="date-column">Date and Time</th>
          <th className="theme-column">Theme</th>
          {past && <th className="ticket-column">Track listing</th>}
          {!past && <th className="ticket-column">Your Tickets</th>}
        </tr>
      </thead>
      <tbody>
        {games.map((game, idx) => (<TableRow game={game} key={idx} past={past} />))}
      </tbody>
    </table>
  );
};
