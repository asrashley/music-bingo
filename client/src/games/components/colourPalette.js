
export const colours = [
  '#FF0000',
  '#008000',
  '#00EFEF',
  '#808000',
  '#BFBF00',
  '#808080',
  '#008080',
  '#00EF00',
  '#C0C0C0',
  '#800000',
  '#0000FF',
  '#000080',
  '#FF00FF',
  '#800080',
];
