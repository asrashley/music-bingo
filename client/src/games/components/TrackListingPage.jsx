import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { initialState } from '../../app/initialState';
import { fetchUserIfNeeded, userIsLoggedIn } from '../../user/userSlice';
import { fetchGamesIfNeeded, fetchDetailIfNeeded, invalidateGameDetail } from '../gamesSlice';
import { LoginDialog } from '../../user/components/LoginDialog';
import { TrackListing } from './TrackListing';
import { ModifyGame } from './ModifyGame';
import routes from '../../routes';

import '../styles/games.scss';

class TrackListingPage extends React.Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    loggedIn: PropTypes.bool.isRequired,
    game: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
  };

  componentDidMount() {
    const { dispatch, gamePk } = this.props;
    dispatch(fetchUserIfNeeded());
    dispatch(fetchGamesIfNeeded());
    dispatch(fetchDetailIfNeeded(gamePk));
  }

  componentDidUpdate(prevProps, prevState) {
    const { dispatch, gamePk } = this.props;
    if (prevProps.user.pk !== this.props.user.pk) {
      dispatch(fetchGamesIfNeeded());
    }
    if (prevProps.game.placeholder !== this.props.game.placeholder) {
      dispatch(fetchDetailIfNeeded(gamePk));
    }
  }

  reloadDetail = (ev) => {
    const { dispatch, gamePk } = this.props;
    ev.preventDefault();
    dispatch(invalidateGameDetail({ gamePk }));
    dispatch(fetchDetailIfNeeded(gamePk));
    return false;
  }

  onDelete = () => {
    const { history } = this.props;
    history.push(`${routes.pastGames}`);
  }

  render() {
    const { dispatch, game, loggedIn, user } = this.props;
    return (
      <div id="track-listing-page" className={loggedIn ? '' : 'modal-open'}  >
        {user.groups.admin === true && <ModifyGame game={game} dispatch={dispatch} onDelete={this.onDelete}
          reload={this.reloadDetail} />}
        <TrackListing game={game} reload={this.reloadDetail} />
        {!loggedIn && <LoginDialog backdrop dispatch={this.props.dispatch} user={user} onSuccess={() => null} />}
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  state = state || initialState;
  const { user } = state;
  const { gamePk } = ownProps.match.params;
  let game = state.games.games[gamePk];
  if (game === undefined) {
    game = {
      title: '',
      pk: -1,
      tracks: [],
      placeholder: true,
    };
  }
  return {
    loggedIn: userIsLoggedIn(state),
    game,
    gamePk,
    user,
  };
};

TrackListingPage = connect(mapStateToProps)(TrackListingPage);

export {
  TrackListingPage
};
