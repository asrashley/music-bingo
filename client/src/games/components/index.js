import { ListGamesPage } from './ListGamesPage';
import { TrackListing } from './TrackListing';
import { TrackListingPage } from './TrackListingPage';
import { ModifyGame } from './ModifyGame';
import { PastGamesPage } from './PastGamesPage';

export {
  ListGamesPage,
  ModifyGame,
  PastGamesPage,
  TrackListing,
  TrackListingPage,
};
