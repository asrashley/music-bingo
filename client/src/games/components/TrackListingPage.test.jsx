import React from 'react';
import { getByText, screen } from '@testing-library/react';
import log from 'loglevel';
import fetchMock from "fetch-mock-jest";

import { createStore } from '../../store/createStore';
import { initialState } from '../../store/initialState';

import { renderWithProviders, installFetchMocks } from '../../testHelpers';
import { formatDuration } from '../../components/DateTime';
import { TrackListingPage, TrackListingPageComponent } from './TrackListingPage';
import gameFixture from '../../fixtures/game/159.json';
import user from '../../fixtures/userState.json';

describe('TrackListingPage component', () => {
  let apiMocks;

  beforeEach(() => {
    apiMocks = installFetchMocks(fetchMock, { loggedIn: true });
  });
  afterEach(() => {
    fetchMock.mockReset();
    log.resetLevel();
    apiMocks = null;
  });

  it('to render track listing for game 18-04-22-2', async () => {
    const history = {
      push: jest.fn()
    };
    const location = {
      params: {
        gameId: "18-04-22-2"
      }
    };
    const { asFragment, store } = renderWithProviders(<TrackListingPage history={history} match={location} />);
    await screen.findByText(`Track listing for Game ${location.params.gameId}`, { exact: false });
    const game = store.getState().games.games[159];
    game.tracks.forEach(track => {
      const tid = `track[${track.pk}]`;
      const dur = formatDuration(track.duration);
      expect(dur).not.toMatch(/N?aN?/);
      const row = screen.getByTestId(tid);
      expect(row).toBeVisible();
      expect(row).toHaveClass(`${game.options.colour_scheme}-theme`);
      getByText(row, track.title);
      getByText(row, track.artist);
      getByText(row, track.album);
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('fetches game details', async () => {
    const store = createStore({
      ...initialState,
      user,
      games: {
        ...initialState.games,
        games: {
          [gameFixture.pk]: {
            ...gameFixture,
            tracks: [],
            ticketOrder: [],
            isFetchingDetail: false,
            invalidDetail: true,
            invalid: false,
          }
        },
        gameIds: {
          [gameFixture.id]: gameFixture.pk,
        },
        pastOrder: [
          gameFixture.pk
        ],
        user: user.pk,
      }
    });
    const history = {
      push: jest.fn()
    };
    const props = {
      dispatch: store.dispatch,
      game: gameFixture,
      user,
      history
    };
    const fetchGameSpy = jest.fn((_, data) => data);
    apiMocks.setResponseModifier('/api/game/159', fetchGameSpy);
    renderWithProviders(<TrackListingPageComponent {...props} />, { store });
    await screen.findByText(`Track listing for Game ${props.game.id}`, { exact: false });
    expect(fetchGameSpy).toHaveBeenCalledTimes(1);
  });
});