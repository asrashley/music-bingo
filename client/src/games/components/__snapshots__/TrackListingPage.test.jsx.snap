// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackListingPage component to render track listing for game 18-04-22-2 1`] = `
<DocumentFragment>
  <div
    class=""
    id="track-listing-page"
  >
    <div
      class="action-panel"
    >
      <button
        class="btn btn-primary ml-2"
      >
        Import Game
      </button>
      <button
        class="btn btn-primary ml-2"
      >
        Export game
      </button>
      <button
        class="btn btn-danger ml-2"
      >
        Delete game
      </button>
    </div>
    <form
      class="modify-game border"
    >
      <button
        class="btn btn-light refresh-icon btn-sm"
      >
        â†»
      </button>
      <div
        class="form-group"
      >
        <label
          for="field-title"
        >
          Title
          <span
            class="required"
          >
            *
          </span>
        </label>
        <input
          class="form-control title  "
          id="field-title"
          name="title"
          placeholder="Title"
          type="text"
        />
        <small
          class="form-text text-muted"
        >
          Title for this round
        </small>
      </div>
      <div
        class="form-group row"
      >
        <div
          class="col-5"
        >
          <label
            for="field-startDate"
          >
            Start Date
            <span
              class="required"
            >
              *
            </span>
          </label>
          <input
            class="form-control start  "
            id="field-startDate"
            name="startDate"
            placeholder="Start Date"
            type="date"
          />
        </div>
        <div
          class="col-5"
        >
          <label
            for="field-startTime"
          >
            Start Time
            <span
              class="required"
            >
              *
            </span>
          </label>
          <input
            class="form-control start  "
            id="field-startTime"
            name="startTime"
            placeholder="Start Time"
            type="time"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <div
          class="col-5"
        >
          <label
            for="field-endDate"
          >
            End Date
            <span
              class="required"
            >
              *
            </span>
          </label>
          <input
            class="form-control end  "
            id="field-endDate"
            name="endDate"
            placeholder="End Date"
            type="date"
          />
        </div>
        <div
          class="col-5"
        >
          <label
            for="field-endTime"
          >
            End Time
            <span
              class="required"
            >
              *
            </span>
          </label>
          <input
            class="form-control end  "
            id="field-endTime"
            name="endTime"
            placeholder="End Time"
            type="time"
          />
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          for="field-colour"
        >
          Colour Scheme
        </label>
        <select
          class="form-control colour  "
          id="field-colour"
          name="colour"
          placeholder="Colour Scheme"
        >
          <option
            value="blue"
          >
            blue
          </option>
          <option
            value="christmas"
          >
            christmas
          </option>
          <option
            value="cyan"
          >
            cyan
          </option>
          <option
            value="green"
          >
            green
          </option>
          <option
            value="grey"
          >
            grey
          </option>
          <option
            value="magenta"
          >
            magenta
          </option>
          <option
            value="orange"
          >
            orange
          </option>
          <option
            value="pink"
          >
            pink
          </option>
          <option
            value="pride"
          >
            pride
          </option>
          <option
            value="purple"
          >
            purple
          </option>
          <option
            value="red"
          >
            red
          </option>
          <option
            value="yellow"
          >
            yellow
          </option>
        </select>
        <small
          class="form-text text-muted"
        >
          Colour scheme for this round
        </small>
      </div>
      <div
        class="form-group"
      >
        <label
          for="field-artist"
        >
          Include Artist
        </label>
        <input
          class="form-control   "
          id="field-artist"
          name="artist"
          placeholder="Include Artist"
          type="checkbox"
        />
      </div>
      <div
        class="clearfix"
      >
        <button
          class="btn btn-success login-button mr-4"
          type="submit"
        >
          Save Changes
        </button>
        <button
          class="btn btn-warning mr-4"
          name="reset"
          type="button"
        >
          Discard Changes
        </button>
      </div>
    </form>
    <div>
      <table
        class="table table-bordered track-listing"
      >
        <thead>
          <tr>
            <th
              class="heading"
              colspan="6"
            >
              Track listing for Game 18-04-22-2: "Various Artists"
            </th>
          </tr>
          <tr>
            <th
              class="number"
            >
              #
            </th>
            <th
              class="start-time"
            >
              Start
            </th>
            <th
              class="title"
            >
              Title
            </th>
            <th
              class="artist"
            >
              Artist
            </th>
            <th
              class="album"
            >
              Album
            </th>
            <th
              class="duration"
            >
              Duration
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="track blue-theme"
            data-testid="track[6597]"
          >
            <td
              class="number"
            >
              1
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Dancing In The Moonlight
            </td>
            <td
              class="artist"
            >
              Toploader
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6598]"
          >
            <td
              class="number"
            >
              2
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Here I Go Again
            </td>
            <td
              class="artist"
            >
              Whitesnake
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6599]"
          >
            <td
              class="number"
            >
              3
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Give It Away
            </td>
            <td
              class="artist"
            >
              Red Hot Chili Peppers
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6600]"
          >
            <td
              class="number"
            >
              4
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              In The Midnight Hour
            </td>
            <td
              class="artist"
            >
              Wilson Pickett
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6601]"
          >
            <td
              class="number"
            >
              5
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              A Message To You Rudy (ft. Rico)
            </td>
            <td
              class="artist"
            >
              The Specials
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6602]"
          >
            <td
              class="number"
            >
              6
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Love Train
            </td>
            <td
              class="artist"
            >
              The O'jays
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6603]"
          >
            <td
              class="number"
            >
              7
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Le Freak
            </td>
            <td
              class="artist"
            >
              Chic
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6604]"
          >
            <td
              class="number"
            >
              8
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Worried About Ray
            </td>
            <td
              class="artist"
            >
              The Hoosiers
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6605]"
          >
            <td
              class="number"
            >
              9
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              (Dont Fear) The Reaper
            </td>
            <td
              class="artist"
            >
              Blue Oyster Cult
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6606]"
          >
            <td
              class="number"
            >
              10
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              December, 1963 (Oh What A Night!)
            </td>
            <td
              class="artist"
            >
              Frankie Valli & The Four Seasons
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6607]"
          >
            <td
              class="number"
            >
              11
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Brown Eyed Girl
            </td>
            <td
              class="artist"
            >
              Van Morrison
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6608]"
          >
            <td
              class="number"
            >
              12
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Didn't I (Blow Your Mind This Time)
            </td>
            <td
              class="artist"
            >
              The Delfonics
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6609]"
          >
            <td
              class="number"
            >
              13
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Break On Through
            </td>
            <td
              class="artist"
            >
              The Doors
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6610]"
          >
            <td
              class="number"
            >
              14
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Pick Up The Pieces
            </td>
            <td
              class="artist"
            >
              Average White Band
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6611]"
          >
            <td
              class="number"
            >
              15
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Ride A Wild Horse
            </td>
            <td
              class="artist"
            >
              Dee Clark
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6612]"
          >
            <td
              class="number"
            >
              16
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Ferry Cross The Mersey (Mono)
            </td>
            <td
              class="artist"
            >
              Gerry & The Pacemakers
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6613]"
          >
            <td
              class="number"
            >
              17
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              September
            </td>
            <td
              class="artist"
            >
              Earth Wind & Fire
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6614]"
          >
            <td
              class="number"
            >
              18
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Crazy Chick
            </td>
            <td
              class="artist"
            >
              Charlotte Church
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6615]"
          >
            <td
              class="number"
            >
              19
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Black Betty
            </td>
            <td
              class="artist"
            >
              Ram Jam
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6616]"
          >
            <td
              class="number"
            >
              20
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              School's Out
            </td>
            <td
              class="artist"
            >
              Alice Cooper
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6617]"
          >
            <td
              class="number"
            >
              21
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              I'm Coming Out
            </td>
            <td
              class="artist"
            >
              Diana Ross
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6618]"
          >
            <td
              class="number"
            >
              22
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Pretty In Pink
            </td>
            <td
              class="artist"
            >
              The Psychedelic Furs
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6619]"
          >
            <td
              class="number"
            >
              23
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              A Groovy Kind Of Love
            </td>
            <td
              class="artist"
            >
              Wayne Fontana & The Mindbenders
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6620]"
          >
            <td
              class="number"
            >
              24
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              (You Got Me) Danglin' On A String
            </td>
            <td
              class="artist"
            >
              Chairmen Of The Board
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6621]"
          >
            <td
              class="number"
            >
              25
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Soul City Walk
            </td>
            <td
              class="artist"
            >
              Archie Bell And The Drells
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6622]"
          >
            <td
              class="number"
            >
              26
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Home Is Where The Heart Is
            </td>
            <td
              class="artist"
            >
              Bobby Womack & The Brotherhood
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6623]"
          >
            <td
              class="number"
            >
              27
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              (Sittin' On) The Dock Of The Bay
            </td>
            <td
              class="artist"
            >
              Otis Redding
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6624]"
          >
            <td
              class="number"
            >
              28
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Family Affair
            </td>
            <td
              class="artist"
            >
              Sly And The Family Stone
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6625]"
          >
            <td
              class="number"
            >
              29
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Nu-flow
            </td>
            <td
              class="artist"
            >
              Big Brovaz
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
          <tr
            class="track blue-theme"
            data-testid="track[6626]"
          >
            <td
              class="number"
            >
              30
            </td>
            <td
              class="start-time"
            >
              00:00.0
            </td>
            <td
              class="title"
            >
              Love & Pride
            </td>
            <td
              class="artist"
            >
              King
            </td>
            <td
              class="album"
            >
              lost+found
            </td>
            <td
              class="duration"
            >
              00:00.0
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
